README.md
# mapbox-map-image-export

> :warning: This tool is experimental. :warning:

A command-line tool to export a [Mapbox vector map][1], such as one created with [Mapbox Studio][2], to a high-resolution image for printing etc.

Mapbox has a [Maps static API][3] for exporting images, but it is limited to `1280px`, and exports as jpeg, with visible compression artifacts. `mapbox-map-image-export` renders the map in a headless browser and exports by default to uncompressed png. You can adjust the dpi setting for high-quality printing. `192dpi` (the default) is good enough for most print applications, but if you want better quality try `288dpi`. Icons will be 192dpi (retina) due to the way sprites are handled by `mapbox-gl-js`.

The Mapbox [wmts service][4] is another option for creating maps for printing in QGIS or ArcMap, but it does not currently support retina maps, so is limited to printing at around 72dpi, and is also only jpeg right now.

[1]: https://www.mapbox.com/maps/
[2]: https://www.mapbox.com/mapbox-studio/
[3]: https://www.mapbox.com/api-documentation/#static
[4]: https://www.mapbox.com/help/mapbox-arcgis-qgis/

```sh
export MAPBOX_TOKEN=pk.eyJ1IjoidGF0aWFuYSIsImEiOiJjaWs1bzRiZGQwMDdjcHRrc285bTdwcWU5In0.0EWPVHyjaE9jTzNvOiIO-w

export-map mapbox://styles/mapbox/streets-v9 -w=11in -h=8.5in \
  -b=-7.1354,57.9095,-6.1357,58.516 -t=$MAPBOX_TOKEN -o=lewis.png
```

![lewis](lewis.png)

## Install

This tool uses Electron to render the map using [mapbox-gl-js][3], so it is fairly heavy (~100mb). Install with latest [npm](https://www.npmjs.com/).

[5]: https://www.mapbox.com/mapbox-gl-js/api/

```sh
npm install mapbox-map-image-export -g
```

## Usage

```txt
Usage:
  export-map mapboxStyleUrl

Options:
  --bounds, -b      comma-separated bounding box [required] 'minLon,minLat,maxLon,maxLat' eg. '-7.1354,57.9095,-6.1357,58.516'
  --token, -t       Mapbox API token [required]: https://www.mapbox.com/studio/account/tokens/
  --output, -o      image output path, optional, defaults to std.out
  --dpi, -d         dpi of output image, default 192dpi (equivalent to how the map renders on a retina screen)
  --format, -f      output format, "jpg", "webp" or "png" (default)
  --quality, -q     encoding quality for jpg and webp, default 0.9
  --width, -w       output width for printing e.g. 11in or 297mm
  --height, -h      output height for printing
```

Exports a map for a given mapbox style url for the specified bounding box.

If no `--output` is given, the PNG is written to stdout.

## What does 'dpi' mean?

Mapbox Studio and Mapbox GL maps are designed for the screen, as such 'dpi' (dots per inch), a print term, does not have any real meaning. However, the standard screen resolution is equivalent to between 72 and 96dpi (e.g. 96 screen pixels will measure approximately 1 inch on a standard screen). Mapbox GL maps are designed by zoom level, and this tool assumes you want maps to look on paper roughly the same as they look on screen at the same size. `mapbox-map-image-export` assumes a standard resolution mapbox map is 96dpi, and a retina (@2x) map is 192dpi. You can set 288dpi if you like, and this will render as if to an ultra-retina screen with a devicePixelRatio=3. Keep dpi to a multiple of `96` if you want icons to look good.

## Acknowledgements

This is inspired by [extract-streetview](https://github.com/Jam3/extract-streetview) by [mattdesl](https://github.com/mattdesl). It uses [Electron](https://electron.atom.io/) to render the map in a headless version of Chrome.

## License

MIT


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [2.2.0] - 2017-07-06
### Changed
- Switch from using [jam3/devtool](https://github.com/Jam3/devtool) to directly spawning electron

### Added
- Log progress to a single line

### Fixed
- Ensure tiles are loaded before exporting map

### Known Issues
- Export to stdout does not exit process on completion (need to Ctrl-C to exit)

[2.2.0]: https://github.com/digidem/mapbox-map-image-export/compare/v2.1.0...v2.2.0


lib/map_mosaic_stream.js
var Qty = require('js-quantities')
var PNGEncoder = require('png-stream/encoder')
var pumpify = require('pumpify')
var mapboxgl = require('mapbox-gl/dist/mapbox-gl-dev.js')
var mosaic = require('mosaic-image-stream')
var spy = require('through2-spy')

var calcViewport = require('./viewport')
var mapImageStream = require('./map_image_stream')
var getScale = require('./scale')

var DEFAULTS = {
  format: 'image/png',
  quality: 0.9,
  dpi: 192,
  width: '11in',
  height: '8.5in'
}

module.exports = mapMosaicStream

function mapMosaicStream (style, opts) {
  opts = Object.assign({}, DEFAULTS, opts)
  var bbox = parseBboxArg(opts.bbox)

  var token = mapboxgl.accessToken = opts.token || process.env.MAPBOX_TOKEN
  if (!token) {
    throw new Error('you must pass a valid Mapbox public token: https://www.mapbox.com/studio/account/tokens/')
  }

  var pixelRatio = window.devicePixelRatio = opts.dpi / 96
  // Dimensions in absolute pixels
  var pixelDim = [opts.width, opts.height]
    .map(function (d) { return Math.ceil(parseLengthToPixels(d, opts.dpi) / pixelRatio) * pixelRatio })
  // Dimensions in display pixels
  var dpDim = pixelDim.map(function (d) { return d / pixelRatio })

  var mapDiv = document.createElement('div')
  document.body.appendChild(mapDiv)

  var map = new mapboxgl.Map({
    container: mapDiv,
    style: style
  })

  var viewport = calcViewport.fitDim(map, dpDim, bbox)

  map.jumpTo({
    center: viewport.center,
    zoom: viewport.zoom
  })

  var scale = getScale(map, dpDim[0], parseLengthToMeters(opts.width))
  console.error('SCALE: 1:%d', scale)
  console.error('ZOOM: %d', viewport.zoom)
  var gl = map.painter.gl
  var maxRenderBuf = gl.getParameter(gl.MAX_RENDERBUFFER_SIZE)
  // Above a certain screen pixel size the map does not render correctly
  var maxTileDim = Array(2).fill(Math.min(1500 * pixelRatio, Math.floor(maxRenderBuf / 2 / pixelRatio)))
  map.remove()

  var tiles = getTileStreams(viewport, dpDim, maxTileDim, style)

  var mapMosaic = mosaic(tiles, pixelDim[1])

  var pngStream = new PNGEncoder({
    width: pixelDim[0],
    height: pixelDim[1],
    colorSpace: 'rgba'
  })
  var total = pixelDim[0] * pixelDim[1] * 4
  var complete = 0
  var out = pumpify(mapMosaic, spy(onChunk), pngStream)
  pngStream.on('format', function (f) {
    out.emit('format', f)
  })
  function onChunk (chunk) {
    complete += chunk.length
    out.emit('progress', complete / total, total)
  }
  return out
}

function getTileStreams (viewport, dim, maxTileDim, style) {
  var ne = viewport.bounds.getNorthEast()
  var sw = viewport.bounds.getSouthWest()
  var pixels2lng = dim[0] / (ne.lng - sw.lng)
  var pixels2lat = dim[1] / (ne.lat - sw.lat)
  var maxTileDimDeg = [maxTileDim[0] / pixels2lng, maxTileDim[1] / pixels2lat]

  var cols = Math.ceil(dim[0] / maxTileDim[0])
  var rows = Math.ceil(dim[1] / maxTileDim[1])

  return Array(cols).fill(1).map(function (_, i) {
    var row = 0
    var mapDiv = document.createElement('div')
    mapDiv.className = 'map'
    document.body.appendChild(mapDiv)
    var map = new mapboxgl.Map({
      container: mapDiv,
      style: style
      // preserveDrawingBuffer: true
    })
    return function createStream (cb) {
      if (row >= rows) return cb(null, null)
      var w = sw.lng + (i * maxTileDimDeg[0])
      var s = Math.max(sw.lat, ne.lat - ((row + 1) * maxTileDimDeg[1]))
      var e = Math.min(ne.lng, sw.lng + ((i + 1) * maxTileDimDeg[0]))
      var n = ne.lat - (row * maxTileDimDeg[1])
      var width = (i + 1) * maxTileDim[0] <= dim[0] ? maxTileDim[0] : dim[0] % maxTileDim[0]
      var height = (row + 1) * maxTileDim[1] <= dim[1] ? maxTileDim[1] : dim[1] % maxTileDim[1]
      row++
      if (map.loaded()) {
        cb(null, mapImageStream(map, [width, height], [w, s, e, n]))
      } else {
        map.on('load', function () {
          cb(null, mapImageStream(map, [width, height], [w, s, e, n]))
        })
      }
    }
  })
}

function parseBboxArg (str) {
  var b = str.split(/,|\s/).map(parseFloat)
  var invalid = b.length !== 4 ||
    b[0] >= b[2] ||
    b[1] >= b[3] ||
    b[0] < -180 ||
    b[2] > 180 ||
    b[1] < -90 ||
    b[3] > 90
  if (invalid) throw new Error('Must pass a valid bounding box')
  return b
}

function parseLengthToMeters (length) {
  var qty = new Qty(length)
  // TODO: This should assume pixels and 96 dpi?
  // as it is this will give a meaningless scale for unitless map sizes
  if (qty.isUnitless()) return length
  return parseFloat(qty.to('meters').toString())
}

function parseLengthToPixels (length, dpi) {
  dpi = dpi || 96
  var qty = new Qty(length)
  if (qty.isUnitless()) return length
  if (qty.kind() !== 'length') throw new Error('Invalid units')
  // Yikes, can't find method to get a unitless number back!
  // TODO: Use a better unit conversion library
  return Math.ceil(parseFloat(qty.to('in').toString()) * dpi)
}

// function parseFormat (format) {
//   switch (format) {
//     case 'jpg':
//     case 'jpeg':
//     case 'image/jpg':
//     case 'image/jpeg':
//       return 'image/jpeg'
//     case 'webp':
//     case 'image/webp':
//       return 'image/webp'
//     default:
//       return 'image/png'
//   }
// }
